### PR TITLE
feat: doc-deploy-changelog action

### PR DESCRIPTION
**Pull request by klmcadams**
This action creates a new section in the CHANGELOG.md file using the towncrier file fragments located in doc/changelog.d. By default, it updates the CHANGELOG in the release branch corresponding to the tag. For example, if the tag is v0.1.0, the action is going to assume its release branch is "release/0.1". Once it updates the changelog in the release branch, it creates a pull request into main containing the changelog update and the deleted fragment files, deletes the old tag, and recreates it with the changelog update. 

If you add the "package" line under the "tool.towncrier" section and set it equal to the value of the name in "tool.flit.module" ("ansys.<product>.<library>"), towncrier will use the version listed in the pyproject.toml file. Otherwise, it will take the version from the tag: v0.1.0 -> 0.1.0.

The action also handles cherry-pick conflicts. It takes all of "their" changes, which are the changes from the initial changelog update & fragment files being deleted. 

If for some reason you push a tag, v0.2.0 and the release branch, release/0.2, doesn't exist, the action will use the temporary branch to update the changelog & merge the changes into main. This may be an overkill check, so if you think it's unnecessary it can be removed.

### CODE REVIEW

File: doc/source/migrations/index.rst [hunk 1]
```diff
@@ -16,6 +16,11 @@ Development version
   This is useful when you want to hide the vulnerabilities from the logs, but still want to fail the action if vulnerabilities are found.
 - Avoid creating issues by default if vulnerabilities are found in the ``ansys/actions/check-vulnerabilities`` action.
 - Create a changelog fragment file for each pull request using ``towncrier`` in the ``ansys/actions/doc-changelog`` action.
+- Vulnerability advisories can now be uploaded to the PyAnsys Dashboard using the ``ansys/actions/check-vulnerabilities`` action
+  by setting the ``pyansys-dashboard-upload`` input to ``true``, together with the necessary credentials.
+- Generate a new section in ``CHANGELOG.md`` if fragment files exist using ``towncrier`` in the ``ansys/actions/doc-deploy-changelog`` action.
```
Comment (RobPasMue) [inline doc/source/migrations/index.rst:21]:
```suggestion
- Generate a new section in ``CHANGELOG.md`` if fragment files exist using ``towncrier`` in the ``ansys/actions/doc-deploy-changelog`` action.
```

File: doc/source/migrations/index.rst [hunk 2]
```diff
@@ -24,6 +29,8 @@ Development version
 **Migration steps:**
 
 - To set up your repository to use the ``ansys/actions/doc-changelog`` action, see the :ref:`docs_changelog_action_setup`.
+- To upload vulnerability advisories to the PyAnsys Dashboard, see the :ref:`pyansys_dashboard_upload`.
+- To set up your repository to use the ``ansys/actions/doc-deploy-changelog`` action, see the :ref:`doc_deploy_changelog_action_setup`.
```
Comment (RobPasMue) [inline doc/source/migrations/index.rst:33]:
```suggestion
- To set up your repository to use the ``ansys/actions/doc-deploy-changelog`` action, see the :ref:`doc_deploy_changelog_action_setup`.
```

File: doc-deploy-changelog/action.yml [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+          # maint/changelog-update-v0.1.0
+          echo "UPDATE_BRANCH=maint/changelog-update-v${{ env.TAG_VERSION }}" >> $GITHUB_ENV
+          echo "RELEASE_BRANCH_EXISTS=false" >> $GITHUB_ENV
+        fi
+
+    - name: "Create new section in CHANGELOG and reupload tag"
+      shell: bash
+      run: |
+        # Get number of .md files in ${{ env.TOWNCRIER_DIR }}
+        # Figure out how to use towncrier's directory from pyproject.toml
```
Comment (RobPasMue) [inline doc-deploy-changelog/action.yml:152]:
```suggestion
```

File: doc-deploy-changelog/action.yml [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+          # Continue the cherry-pick without a commit message
+          git -c core.editor=true cherry-pick --continue
+        fi
+
+        # Push CHANGELOG.md changes to $pr_branch
+        git push -u origin $pr_branch
+
+        gh pr create --title "Update CHANGELOG for v${{ env.TAG_VERSION }}" --body "Update CHANGELOG for v${{ env.TAG_VERSION }} and remove .md files in ${{ env.TOWNCRIER_DIR }}" --reviewer ${{ github.actor }}
+
```
Comment (RobPasMue) [inline doc-deploy-changelog/action.yml:259]:
```suggestion
        gh pr create --title "chore: update CHANGELOG for v${{ env.TAG_VERSION }}" --body "Update CHANGELOG for v${{ env.TAG_VERSION }} and remove .md files in ${{ env.TOWNCRIER_DIR }}" --reviewer ${{ github.actor }}

```

**Review by RobPasMue — APPROVED** [review]
Awesome job @klmcadams - can't wait to try it out. Just left a few comments (mostly editorial). Logic seems to work fine.

File: doc/source/migrations/doc-deploy-changelog-setup.rst [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      name: "Update CHANGELOG for new tag"
+      if: github.event_name == 'push' && contains(github.ref, 'refs/tags')
+      runs-on: ubuntu-latest
+      permissions:
+        contents: write
+        pull-requests: write
+      steps:
+        - uses: ansys/actions/doc-deploy-changelog@{{ version }}
+          with:
+            token: ${{ secrets.PYANSYS_CI_BOT_TOKEN }}
```
Comment (RobPasMue) [inline doc/source/migrations/doc-deploy-changelog-setup.rst:28]:
```suggestion
            token: ${{ '{{ secrets.PYANSYS_CI_BOT_TOKEN }}' }}
```

File: doc/source/migrations/doc-deploy-changelog-setup.rst [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    release:
+      name: Release project
+      if: github.event_name == 'push' && contains(github.ref, 'refs/tags')
+      needs: [update-changelog]
+      runs-on: ubuntu-latest
+      steps:
+        - name: Release to the public PyPI repository
+          uses: ansys/actions/release-pypi-public@v5
+          with:
+            library-name: ${{ env.PACKAGE_NAME }}
```
Comment (RobPasMue) [inline doc/source/migrations/doc-deploy-changelog-setup.rst:41]:
```suggestion
            library-name: ${{ '{{ env.PACKAGE_NAME }}' }}
```

File: doc/source/migrations/doc-deploy-changelog-setup.rst [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            twine-username: "__token__"
+            twine-token: ${{ secrets.PYPI_TOKEN }}
```
Comment (RobPasMue) [inline doc/source/migrations/doc-deploy-changelog-setup.rst:43]:
```suggestion
            twine-token: ${{ '{{ secrets.PYPI_TOKEN }}' }}
```

**Review by RobPasMue — COMMENTED** [review]
This should fix the doc-build errors

**Comment by RobPasMue** [comment]
@jorgepiloto - this shouldn't be blocked I believe. We could merge it to main like we've done with others. Especially now that we are close to v6.

File: doc-changelog/action.yml [hunk 8]
```diff
@@ -82,7 +82,7 @@ runs:
     - name: "Install towncrier"
       shell: bash
       run: |
-        python -m pip install --upgrade pip towncrier
```
Comment (jorgepiloto) [inline doc-changelog/action.yml:85]: Could we pass this as an input with default value `23.11.0`?

Comment (klmcadams) [reply doc-changelog/action.yml:85]: Should I do the same for the doc-deploy-changelog action too?

Comment (RobPasMue) [reply doc-changelog/action.yml:85]: Yep please do

File: doc-changelog/action.yml [hunk 9]
```diff
@@ -82,7 +89,7 @@ runs:
...
+        python -m pip install --upgrade pip towncrier==${{ inputs.towncrier-version }}
```
Comment (klmcadams) [inline doc-changelog/action.yml:92]: @jorgepiloto Is this what you meant?

Comment (RobPasMue) [reply doc-changelog/action.yml:92]: Yep. That's it

**Pull request merged**